overlord/snapstate/component.go

```diff
@@ -0,0 +1,227 @@
+
+/*
+ * Copyright (C) 2023 Canonical Ltd
+ *
+ * This program is free software: you can redistribute it and/or modify
+ * it under the terms of the GNU General Public License version 3 as
+ * published by the Free Software Foundation.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ *
+ */
+
+package snapstate
+
+import (
+	"errors"
+	"fmt"
+	"os"
+
+	"github.com/snapcore/snapd/i18n"
+	"github.com/snapcore/snapd/overlord/snapstate/backend"
+	"github.com/snapcore/snapd/overlord/state"
+	"github.com/snapcore/snapd/snap"
+)
+
+// InstallComponentPath returns a set of tasks for installing a snap component
+// from a file path.
+//
+// Note that the state must be locked by the caller. The provided SideInfo can
+// contain just a name which results in local sideloading of the component, or
+// full metadata in which case the component will appear as installed from the
+// store.
+func InstallComponentPath(st *state.State, csi *snap.ComponentSideInfo, si *snap.SideInfo,
+	path, snapInstance, channel string, flags Flags) (*state.TaskSet, error) {
+	if si.RealName == "" {
+		return nil, fmt.Errorf(
+			"internal error: snap name to install component %q not provided",
+			path)
+	}
+
+	if snapInstance == "" {
+		snapInstance = si.RealName
+	}
+
+	deviceCtx, err := DeviceCtxFromState(st, nil)
+	if err != nil {
+		return nil, err
+	}
+
+	var snapst SnapState
+	err = Get(st, snapInstance, &snapst)
+	if err != nil && !errors.Is(err, state.ErrNoState) {
+		return nil, err
+	}
+
+	channel, err = resolveChannel(snapInstance, snapst.TrackingChannel, channel,
+		deviceCtx)
+	if err != nil {
+		return nil, err
+	}
+
+	// Read ComponentInfo
+	compInfo, _, err := backend.OpenComponentFile(path)
+	if err != nil {
+		return nil, err
+	}
+
+	// Check snap name matches
+	if compInfo.Component.SnapName != si.RealName {
+		return nil, fmt.Errorf(
+			"component snap name %q does not match real snap name %q",
+			compInfo.Component.SnapName, si.RealName)
+	}
+
+	// Read snap.Info from the installed snap
+	info, err := readInfo(snapInstance, si, withAuxStoreInfo)
+	if err != nil {
+		return nil, err
+	}
+
+	providerContentAttrs := defaultProviderContentAttrs(st, info, nil)
+	snapsup := &SnapSetup{
+		Base:               info.Base,
+		Prereq:             getKeys(providerContentAttrs),
+		PrereqContentAttrs: providerContentAttrs,
+		SideInfo:           si,
+		SnapPath:           path,
+		Channel:            channel,
+		Flags:              flags.ForSnapSetup(),
+		Type:               info.Type(),
+		Version:            info.Version,
+		PlugsOnly:          len(info.Slots) == 0,
+		InstanceKey:        info.InstanceKey,
```

Comment:
this probably can set just a subset of things? the closest operation would be enable I suppose?